### PR TITLE
idempotent portmap teardown 

### PR DESCRIPTION
**Pull request by aojea**
It turns out that the portmap plugin is not idempotent if its executed in parallel.
The errors are caused due to a race of different iptables executions deleting the chains.
This patch adds a mutex based in a file lock that solves the concurrency problems.

xref: https://github.com/containernetworking/plugins/issues/418

### CODE REVIEW

File: plugins/meta/portmap/chain.go [hunk 1]
```diff
@@ -17,12 +17,16 @@ package main
...
 	"strings"
+	"time"
 
 	"github.com/containernetworking/plugins/pkg/utils"
 	"github.com/coreos/go-iptables/iptables"
+	"github.com/juju/fslock"
 	"github.com/mattn/go-shellwords"
 )
 
+const lockfile = "/tmp/portmap.lock"
```
Comment (aojea) [inline plugins/meta/portmap/chain.go]: is this the right approach?

**Comment by mars1024** [comment]
Related PR #408

**Comment by aojea** [comment]
> Related PR #408

I've added a test to the PR that verifies the idempotency if there are concurrent executions.
I've tried the approach of the mentioned PR but was still flaky and failed constantly without adding a sleep to the test between goroutines. 
The external lock guarantees the atomicity of the operation, as you can see with 10 goroutines there are no errors tearing down the portmap rules

**Comment by mars1024** [comment]
> > Related PR #408
> 
> I've added a test to the PR that verifies the idempotency if there are concurrent executions.
> I've tried the approach of the mentioned PR but was still flaky and failed constantly without adding a sleep to the test between goroutines.
> The external lock guarantees the atomicity of the operation, as you can see with 10 goroutines there are no errors tearing down the portmap rules

Do you have the old PR branch which I can take a look at?
If we want to make a function idempotent, we must make sure every step of this function is idempotent, just like #408, maybe more error handlers are needed.

IMO, global locker should be our last choice, but I believe it will work.

**Comment by aojea** [comment]
@mars1024 I've create a new PR so you can compare https://github.com/containernetworking/plugins/pull/421

**Pull request closed without merging**